DOCSP-37543 - Retrieve Data (#37)
mongodb/docs-pymongo: source/read/retrieve.txt

.. _pymongo-retrieve:

=============
Retrieve Data
=============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: code examples, read, search, cursor

Overview
--------

In this guide, you can learn how to use {+driver-long+} to retrieve
data from a MongoDB collection by using read operations. You can call the

> ``find()`` method to retrieve documents that match a set of criteria.

Reviewer: Since this is an overview for all of retrieve, I think we should mention both find methods that get talked about througout
```suggestion
``find()`` or the ``find_one()`` methods to retrieve documents that match a set of criteria.
```

Author: yep. tried to do that throughout but overlooked this one

   section of this guide.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_restaurants.restaurants`` collection
from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the :ref:`<pymongo-quickstart>`.

.. _pymongo-retrieve-find:

Find Documents
--------------

{+driver-short+} includes two methods for retrieving documents from a collection:
``find_one()`` and ``find()``.
These methods take a **query filter** and return one or more matching documents.
A query filter is an object that specifies the documents you want to retrieve in
your query. 

.. TODO: To learn more about query filters, see :ref:`pymongo-specify-query`.

.. _pymongo-retrieve-find-one:

Find One Document
~~~~~~~~~~~~~~~~~

To find a single document in a collection, call the ``find_one()`` method and pass a query
filter that specifies the criteria of the document you want to find.
If more than one document matches the query
filter, this method returns the *first* matching document from the retrieved
results as a Python dictionary. If no documents match the query filter, the method returns
``None``.

> It is useful when you know there is only one matching document, or are only interested in the first match.  

Reviewer: ```suggestion
The ``find_one()`` method is useful when you know there is only one matching document, or are only interested in the first match.
```

The following examples use the ``find_one()`` method to find the first document where
the ``"cuisine"`` field has the value ``"Bakery"``. Select the corresponding tab to see
an example in the Python shell or a ``.py`` file. 

.. tabs::

   .. tab:: Shell
      :tabid: find-one-shell

      .. code-block:: python
         :copyable: true

         >>> sample_restaurants.restaurants.find_one({"cuisine": "Bakery"})

   .. tab:: File

>       :tabid: find-one-file

Reviewer: I think we should show the "File" tab first in these pages to make them more Atlas-centric, since most of our guidance towards using Atlas is doing it in a code file.

Author: makes sense

      .. code-block:: python
         :copyable: true

         restaurant = sample_restaurants.restaurants.find_one({"cuisine": "Bakery"})

.. tip:: Sort Order

   The ``find_one()`` method returns the first document in 
   :manual:`natural order </reference/glossary/#std-term-natural-order>`
   on disk if no sort criteria is specified.

.. To learn more about sorting, see the TODO: sorting page.

.. TODO: To see a full example of using the ``find_one()`` method to find a single document, see
   :ref:`pymongo-retrieve-additional-information`.

.. _pymongo-retrieve-find-multiple:

Find Multiple Documents
~~~~~~~~~~~~~~~~~~~~~~~

To find multiple documents in a collection, pass a query filter to the ``find()``
method that specifies the criteria of the documents you want to retrieve.

You can use a **cursor** to iterate over the documents returned by the ``find()``
method. A cursor is a mechanism that allows an application to iterate over database 
results while holding only a subset of them in memory at a given time. Cursors

> are useful when your ``find()`` method returns a large amount of documents.

Reviewer: S: I think this paragraph should be moved under the first code example (on line 134 where it says: "You can iterate over the documents..."). That way it separates using the find method from iterating over the results of the find method, and also moves the first mention of a cursor closer to the first example of using one.

Author: great idea

The following examples use the ``find()`` method to find all documents where
the ``"cuisine"`` field has the value ``"Spanish"``. Select the corresponding tab to see
an example in the Python shell or a ``.py`` file. 

.. tabs::

   .. tab:: Shell
      :tabid: find-multiple-shell

      .. code-block:: python
         :copyable: true

         >>> sample_restaurants.restaurants.find({"cuisine": "Spanish"})

   .. tab:: File
      :tabid: find-multiple-file

      .. code-block:: python
         :copyable: true

>          cursor = sample_restaurants.restaurants.find({"hello": "Spanish"})

Reviewer: ```suggestion
         cursor = sample_restaurants.restaurants.find({"cuisine": "Spanish"})
```

You can iterate over the documents in a cursor by using a ``for-in`` loop, as shown in
the following examples:

.. tabs::

   .. tab:: Shell
      :tabid: iterate-shell

      .. code-block:: python
         :copyable: true

         >>> for restaurant in sample_restaurants.restaurants.find({"cuisine": "Spanish"}):

   .. tab:: File
      :tabid: iterate-file

      .. code-block:: python
         :copyable: true

>          cursor = sample_restaurants.restaurants.find({"hello": "Spanish"})

Reviewer: ```suggestion
         cursor = sample_restaurants.restaurants.find({"cuisine": "Spanish"})
```

You can modify the behavior of the ``find()`` and ``find_one()`` methods by passing
named arguments to them. The following table describes the commonly used arguments:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Argument
     - Description

   * - ``batch_size`` 
     - | Limits the number of documents to hold in a cursor at a given time.

   * - ``collation`` 
     - | An instance of the ``Collation`` class that sets the collation options. 

   * - ``comment`` 
     - | A string to attach to the query. This can help you trace and interpret the
         operation in the server logs and in profile data. To learn more about query comments, 
         see the :manual:`$comment </reference/operator/query/comment/>` page.
   
   * - ``hint`` 
     - | The index to use for the query.

   * - ``max_time_ms`` 
     - | The maximum execution time on the server for this operation. If this time is
         exceeded, {+driver-short+} aborts the operation and raises an ``ExecutionTimeout``.

For a full list of available arguments, see the
`API documentation <{+api-root+}pymongo/collection.html#pymongo.collection.Collection.find>`__
for the  ``find() method``.

> .. TODO: .. instruqt lab

Reviewer: Is this Instruqt lab for the Modify Find Behavior section?

Author: No, I think it will be just for a Find.

Reviewer: Okay makes sense. I think we might also want to add a TODO to add some code examples here of using the options in the "Modify" table. Unless that's going to go on one of the more in depth pages?

Find Documents
--------------

{+driver-short+} includes two methods for retrieving documents from a collection:
``find_one()`` and ``find()``.
These methods take a **query filter** and return one or more matching documents.
A query filter is an object that specifies the documents you want to retrieve in
your query. 

.. TODO: To learn more about query filters, see :ref:`pymongo-specify-query`.

.. _pymongo-retrieve-find-one:

Find One Document
~~~~~~~~~~~~~~~~~

To find a single document in a collection, call the ``find_one()`` method and pass a query
filter that specifies the criteria of the document you want to find.
If more than one document matches the query
filter, this method returns the *first* matching document from the retrieved
results as a Python dictionary. If no documents match the query filter, the method returns
``None``.

.. tip::
   
   The ``find_one()`` method is useful when you know there's only one matching document,
   or you're only interested in the first match.  

The following examples use the ``find_one()`` method to find the first document where
the ``"cuisine"`` field has the value ``"Bakery"``. Select the corresponding tab to see

> an example in the Python shell or a ``.py`` file. 

Reviewer: Probably a bit nitpicky, but I noticed these are out of order from the tabs now that they're switched. (applies to all)

```suggestion
an example in a ``.py`` file or the Python shell. 
```

The following examples use the ``find()`` method to find all documents where
the ``"cuisine"`` field has the value ``"Spanish"``. Select the corresponding tab to see
an example in the Python shell or a ``.py`` file. 

.. tabs::

   .. tab:: File
      :tabid: find-multiple-file

      .. code-block:: python
         :copyable: true

         cursor = sample_restaurants.restaurants.find({"cuisine": "Spanish"})

   .. tab:: Shell
      :tabid: find-multiple-shell

      .. code-block:: python
         :copyable: true

         >>> sample_restaurants.restaurants.find({"cuisine": "Spanish"})

You can use a **cursor** to iterate over the documents returned by the ``find()``
method. A cursor is a mechanism that allows an application to iterate over database 
results while holding only a subset of them in memory at a given time. Cursors
are useful when your ``find()`` method returns a large amount of documents.

You can iterate over the documents in a cursor by using a ``for-in`` loop, as shown in
the following examples:

.. tabs::

   .. tab:: File
      :tabid: iterate-file

      .. code-block:: python
         :copyable: true

>          cursor = sample_restaurants.restaurants.find({"cuisine": "Spanish"})
>          for restaurant in cursor:

Reviewer: Might be useful to add an elipses in the code examples here so beginners know there needs to be more after this block. (Alternatively we could just show a print line or something) Applies to shell example too

```suggestion
         cursor = sample_restaurants.restaurants.find({"cuisine": "Spanish"})
         for restaurant in cursor:
         ...
```

.. _pymongo-retrieve:

=============
Retrieve Data
=============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: code examples, read, search, cursor

Overview
--------

In this guide, you can learn how to use {+driver-long+} to retrieve
data from a MongoDB collection by using read operations. You can call the
``find()`` or ``find_one()`` method to retrieve documents that match a set of criteria.

.. TODO: .. tip:: Interactive Lab
   You can complete a short interactive lesson that demonstrates how to
   retrieve data by using the ``Find()`` method in an in-browser coding
   experience. To learn more, see the :ref:`pymongo-retrieve-instruqt-lab`
   section of this guide.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_restaurants.restaurants`` collection
from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a

> free MongoDB Atlas cluster and load the sample datasets, see the :ref:`<pymongo-get-started>`.

Author: Link broken until a later PR is merged
